# Day 12: NumPy Aggregations and Statistics

**Dataset:** `Iris_Dataset.csv`

This notebook uses NumPy functions to calculate sum, mean, median, minimum, maximum, standard
deviation, and variance, using real values pulled directly from the Iris dataset.

## 1. Import Libraries and Load the Dataset

In [ ]:
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv("../datasets/Iris_Dataset.csv")

# Preview the data
df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


**Create a 1D array from one column, and a 2D array from all four numeric feature columns**

In [ ]:
petal_length = np.array(df["petal length (cm)"])

feature_columns = ["sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)"]
features = np.array(df[feature_columns])

print("1D array shape:", petal_length.shape)
print("2D array shape:", features.shape)

1D array shape: (150,)
2D array shape: (150, 4)


## 2. Examples of Aggregation Functions

Each of the functions below reduces an array down to a single summary number.

**1. `sum()` — total of all values**

In [ ]:
print(petal_length.sum())

563.8


**2. `mean()` — average value**

In [ ]:
print(petal_length.mean())

3.758666666666666


**3. `np.median()` — the middle value when sorted**

In [ ]:
print(np.median(petal_length))

4.35


**4. `min()` — smallest value**

In [ ]:
print(petal_length.min())

1.0


**5. `max()` — largest value**

In [ ]:
print(petal_length.max())

6.9


**6. `std()` — standard deviation (how spread out the values are)**

In [ ]:
print(petal_length.std())

1.7585291834055212


**7. `var()` — variance (the standard deviation, squared)**

In [ ]:
print(petal_length.var())

3.092424888888889


**8. `np.ptp()` — the range (max minus min), and `np.percentile()` — a specific percentile**

In [ ]:
print("Range (ptp):", np.ptp(petal_length))
print("25th percentile:", np.percentile(petal_length, 25))
print("75th percentile:", np.percentile(petal_length, 75))

Range (ptp): 5.9
25th percentile: 1.6
75th percentile: 5.1


## 3. Experimenting with `axis=0` and `axis=1`

On the 2D `features` array (150 rows × 4 columns), `axis=0` aggregates **down each column**
(one result per feature), while `axis=1` aggregates **across each row** (one result per
flower).

**`axis=0` — mean of each column (one value per feature, across all 150 flowers)**

In [ ]:
column_means = features.mean(axis=0)
print("Column means [sepal length, sepal width, petal length, petal width]:")
print(column_means)

Column means [sepal length, sepal width, petal length, petal width]:
[5.84333333 3.054      3.75866667 1.19866667]


**`axis=1` — mean of each row (one value per flower, across its 4 features)**

In [ ]:
row_means = features.mean(axis=1)
print("First 10 row means (average of that flower's 4 measurements):")
print(row_means[:10])

First 10 row means (average of that flower's 4 measurements):
[2.55  2.375 2.35  2.35  2.55  2.85  2.425 2.525 2.225 2.4  ]


**Comparing the shapes returned by each axis**

In [ ]:
print("features.shape:", features.shape)
print("mean(axis=0).shape:", features.mean(axis=0).shape)   # one value per column -> shape (4,)
print("mean(axis=1).shape:", features.mean(axis=1).shape)   # one value per row -> shape (150,)

features.shape: (150, 4)
mean(axis=0).shape: (4,)
mean(axis=1).shape: (150,)


## 4. Statistical Analysis of the Dataset

A full statistical summary of all four numeric columns at once, using `axis=0`.

In [ ]:
summary = {
    "mean": features.mean(axis=0),
    "median": np.median(features, axis=0),
    "std": features.std(axis=0),
    "var": features.var(axis=0),
    "min": features.min(axis=0),
    "max": features.max(axis=0)
}

summary_df = pd.DataFrame(summary, index=feature_columns)
summary_df

,mean,median,std,var,min,max
sepal length (cm),5.843333,5.80,0.825301,0.681122,4.3,7.9
sepal width (cm),3.054000,3.00,0.432147,0.186751,2.0,4.4
petal length (cm),3.758667,4.35,1.758529,3.092425,1.0,6.9
petal width (cm),1.198667,1.30,0.760613,0.578532,0.1,2.5


## 5. Comparing Results Across Different Columns

In [ ]:
stds = features.std(axis=0)
for column_name, std_value in zip(feature_columns, stds):
    print(f"{column_name}: std = {std_value:.3f}")

most_variable_index = np.argmax(stds)
least_variable_index = np.argmin(stds)
print(f"\nMost variable feature: {feature_columns[most_variable_index]}")
print(f"Least variable feature: {feature_columns[least_variable_index]}")

sepal length (cm): std = 0.825
sepal width (cm): std = 0.432
petal length (cm): std = 1.759
petal width (cm): std = 0.761

Most variable feature: petal length (cm)
Least variable feature: sepal width (cm)


## 6. Short Interpretation of the Results

Based on the statistics computed above:

- **Petal length** has by far the highest standard deviation (**≈1.76 cm**) and variance
  (**≈3.09**) of the four features, meaning petal length varies the most across the 150
  flowers in this dataset (ranges all the way from **1.0 cm to 6.9 cm**).
- **Sepal width** has the lowest standard deviation (**≈0.43 cm**) and variance (**≈0.19**),
  meaning it's the most consistent measurement across flowers, staying fairly close to its
  mean of **≈3.05 cm**.
- **Petal width** also has a fairly low mean (**≈1.20 cm**) but a moderate spread relative to
  that mean, suggesting, like petal length, it varies more between individual flowers than
  the sepal measurements do.
- Since the dataset contains three different species of iris, this spread in the petal
  measurements (compared to the more consistent sepal measurements) hints that petal length
  and petal width are likely more useful for **telling the three species apart** than the
  sepal measurements are.

## Outcome

This notebook practiced eight NumPy aggregation functions, namely, `sum()`, `mean()`, `np.median()`,
`min()`, `max()`, `std()`, `var()`, and `np.ptp()`/`np.percentile()`, using real values from
the `Iris_Dataset.csv` file. I experimented with `axis=0` (aggregating down each column, to get
one result per feature) and `axis=1` (aggregating across each row, to get one result per
flower) on the same 2D array, and confirmed the difference by checking the shape each axis
setting produces. Building a full statistical summary table across all four feature columns,
then comparing their standard deviations directly, showed that petal length is by far the most
variable measurement in the dataset while sepal width is the most consistent. This is a useful signal
about which features carry the most distinguishing information between the flower species.